### PR TITLE
Recipeitem to abstract

### PR DESCRIPTION
**Pull request by ezrarp**
this should be reviewed, had some issues with the ingredients class not showing up on my branch but it works if that is resolved,

### CODE REVIEW

File: src/main/java/Entities/Implementations/AbstractRecipeItem.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    //public AbstractRecipeItem(Ingredient ingredient, float amount) {
+    //   this.ingredient = ingredient;
+    // this.amount = amount;
+    //}
+    /** Implement a ItemImpl, giving the ingredient,
+     * quantity of this ingredient, and if the ingredient is optional.
+     * @param ingredient  The ingredient
+     * @param amount     The quantity of this ingredient
+     * @param optional  Is the ingredient optional?
+     */
```
Comment (DerekCresswell) [inline src/main/java/Entities/Implementations/AbstractRecipeItem.java:25]: Why is this constructor commented out?

These doc comments are poor. This is not about implementing a `ItemImpl`. The comment should not describe the obvious when possible. It should be plain english about the use case. A feasible thing to say would be this assumes the item is not optional since this is a choice made by the class when using that constructor.

File: src/main/java/Entities/Implementations/AbstractRecipeItem.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package Entities.Implementations;
+
+import Entities.Ingredient;
+import EntityInterfaces.RecipeItem;
+
+public abstract class AbstractRecipeItem implements RecipeItem {
+    public Ingredient ingredient;
+    public float amount;
+    public boolean optional = false;
```
Comment (DerekCresswell) [inline src/main/java/Entities/Implementations/AbstractRecipeItem.java:9]: These should be protected not public. We do not want outside classes to access these.

File: src/main/java/Entities/Implementations/QuantityRecipeItem.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package Entities.Implementations;
+
+import Entities.Implementations.AbstractRecipeItem;
+import Entities.Ingredient;
+
+public class QuantityRecipeItem extends AbstractRecipeItem {
+    /**
+     * Implement a ItemImpl, giving the ingredient,
+     * quantity of this ingredient, and if the ingredient is optional.
```
Comment (DerekCresswell) [inline src/main/java/Entities/Implementations/QuantityRecipeItem.java:9]: Ditto, and for the rest too.

File: src/main/java/EntityInterfaces/RecipeItem.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+     *
+     * returns a float of the quantity needed for the recipe.
+     * @return returns a float of the quantity needed for the recipe.
+     *
+     * displays the recipeitem in string form.
+     * @return a string of the recipeitem.
+     */
+    Ingredient ingredient();
+    float quantity();
+    boolean optional();
```
Comment (DerekCresswell) [inline src/main/java/EntityInterfaces/RecipeItem.java:18]: Don't put comments together like this. Make one for each.

File: src/test/java/Entities/QuantityRecipeItemTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        IngredientImpl ingredient = new IngredientImpl("lemon", Collections.emptyList());
+        VolumetricRecipeItem quantiyitem = new VolumetricRecipeItem(ingredient, 0.5f, false);
+
+        Assertions.assertEquals("0.5 of a lemon", quantiyitem.display());
+
+    }
+
+    @Test
+    public void testDisplayLargeQuantity() {
+        IngredientImpl ingredient = new IngredientImpl("lemon", Collections.emptyList());
```
Comment (DerekCresswell) [inline src/test/java/Entities/QuantityRecipeItemTest.java:31]: The convention we are using for this repo is that each public method gets one method, in this case `testDisplay` which should contain each of these sequentially.

Comment (DerekCresswell) [reply src/test/java/Entities/QuantityRecipeItemTest.java:31]: Also make sure we test all cases, such as correctly using `an` over `a`.

File: src/test/java/Matcher/TagMatcherTest.java [hunk 6]
```diff
@@ -63,7 +62,7 @@ public void testMatchesFalse() {
         RecipeItemImpl item = new RecipeItemImpl(ingredient1, 15);
         List<RecipeItem> recipeItems = new ArrayList<>();
         recipeItems.add(item);
-        RecipeImpl recipe = new RecipeImpl("name", "description", Collections.singletonList("instructions"), recipeItems);
+        RecipeImpl recipe = new RecipeImpl("name", "description", "instructions", recipeItems);
```
Comment (DerekCresswell) [inline src/test/java/Matcher/TagMatcherTest.java:65]: Why were these changed? Instructions should be a list. Put them back.

File: src/main/java/Entities/Implementations/VolumetricRecipeItem.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+     * @param amount     The quantity of this ingredient
+     * @param optional   Is the ingredient optional?
+     */
+    public VolumetricRecipeItem(Ingredient ingredient, float amount, boolean optional) {
+        super(ingredient, amount, optional);
+    }
+
+    public String display(){
+        String stringquantity = Float.toString(this.quantity());
+      return stringquantity + "ml of" + this.ingredient;
```
Comment (DerekCresswell) [inline src/main/java/Entities/Implementations/VolumetricRecipeItem.java:22]: Indentation is weird.

File: src/main/java/Entities/Implementations/QuantityRecipeItem.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        if (this.quantity() < 1) {
+            //checks to see if the first letter of the ingredient name is a vowel.
+            char letter = (this.ingredient.name()).charAt(0);
+
+            if ("AEIOUaeiou".indexOf(letter) != -1) {
+                return stringquantity + "of an" + this.ingredient.name();
+            }
+            return stringquantity + "of a" + this.ingredient.name();
+        }
+        return stringquantity + this.ingredient.name();
```
Comment (DerekCresswell) [inline src/main/java/Entities/Implementations/QuantityRecipeItem.java:35]: This method could be simplified.

Since the body of `<1` is the most  complex, it should not be within an if statement as this nests it further. Try checking `=1`, `>1` and then have the `<1` case not in an if or else statement.

You could also remove the return statement for `of an` and instead declare a string `of a`, if the first letter is a vowel append `n`, and then construct a string using this and return. Cuts out a return which makes it a bit easier to follow. See if it helps.

**Review by DerekCresswell — CHANGES_REQUESTED** [review]

**Comment by DerekCresswell** [comment]
You most also get the tests running and passing. You shouldn't request a review otherwise because we won't merge broken code.

Lastly, these commits are very gross, can you squash them so that there aren't duplicates? Intellij should be able to help with this.

File: src/test/java/Entities/QuantityRecipeItemTest.java (hunk 5 above)
Comment (ezrarp) [reply src/test/java/Entities/QuantityRecipeItemTest.java:31]: do you want me to make a testdisplay class? and then have there be tests for both volume and quantity under the same class. @DerekCresswell

File: src/main/java/EntityInterfaces/RecipeItem.java (hunk 4 above)
Comment (ezrarp) [reply src/main/java/EntityInterfaces/RecipeItem.java:18]: This wasn't my code but can def fix it.

File: src/test/java/Entities/QuantityRecipeItemTest.java (hunk 5 above)
Comment (DerekCresswell) [reply src/test/java/Entities/QuantityRecipeItemTest.java:31]: No. One test class per class, one test method per method.

**Comment by arieldeveloper** [comment]
@DerekCresswell I changed up a bunch of stuff, needed to change some old tests for the placeholder tag matcher class. Plus changed documentation and the display methods were not working properly. Should be good to push.

**Review by DerekCresswell — APPROVED** [review]
